src/common/filemanager.cpp

```diff
@@ -0,0 +1,531 @@
+			{
+				runTaskFromQueue(state, state->read_tasks_queue);
+			}
+			if (state->read_text_tasks_queue.size() == 1)
+			{
+				runTaskFromQueue(state, state->read_text_tasks_queue);
+			}
+			else if (state->write_tasks_queue.size() == 1)
+			{
+				runTaskFromQueue(state, state->write_tasks_queue);
+			}
+			state->state_mutex.unlock();
+
+			{
+			std::unique_lock<std::mutex> lock(state->semaphore_mutex);
+			if (is_eternal_thread)
+				state->semaphore.wait(lock);
+			else
+			{
+				std::cv_status status = state->semaphore.wait_for(lock, THREAD_LIVE_TIMEOUT);
+				if (status == std::cv_status::timeout)
+					state->is_exit.store(true);
+			}
+			}
+		}
+	}
+
+	ReportableWorker* make_worker()
+	{
+		ReportableWorker* new_worker = new ReportableWorker();
+		new_worker->state.has_task = false;
+		if (threads.size() < ETERNAL_THREAD_COUNT)
+			new_worker->state.live_forever = true;
+		else
+			new_worker->state.live_forever = false;
+		threads.push_back(new_worker);
+		return new_worker;
+	}
+
+	void run_worker(ReportableWorker* worker)
+	{
+		packaged_task<void(ReportableWorkerState*)> main_work_task(&FileIoThreadPool::workerFunction);
+		worker->worker = std::thread(std::move(main_work_task), &worker->state);
+	}
+
+	void cleanup_finished_workers()
+	{
+		for (auto iter = threads.begin(); iter != threads.end();)
+		{
+			if ((*iter)->state.is_exit.load())
+			{
+				delete *iter;
+				iter = threads.erase(iter);
+			}
+			else
+				iter++;
+		}
+	}
+
+	template<typename T>
+	future<T> enqueueTask(const path& relative_path, std::function<T(std::filesystem::path)> task_function, std::function<void(ReportableWorkerState*, packaged_task<T()>&)> task_pusher_function)
+	{
+		cleanup_finished_workers();
+
+		packaged_task<T()> task(std::bind(task_function, relative_path));
+
+		voxen::Log::info("enqueue read task to threadpool size {}", threads.size());
+		for (size_t i = 0; i < threads.size(); i++)
+		{
+			ReportableWorker* worker = threads[i];
+			worker->state.state_mutex.lock();
+			if (!worker->state.has_task)
+			{
+				voxen::Log::info("use already created thread #{}", i);
+				future<T> result = task.get_future();
+				task_pusher_function(&worker->state, task);
+				worker->state.has_task = true;
+				worker->state.state_mutex.unlock();
+
+				worker->state.semaphore.notify_one();
+				return result;
+			}
+			else
+				worker->state.state_mutex.unlock();
+		}
+		voxen::Log::info("Make new thread for task");
+
+		// If we here, then no free threads, so construct new for the task
+		// So, make worker, push task and run the worker
+		ReportableWorker* new_worker = make_worker();
+		future<T> result = task.get_future();
+		task_pusher_function(&new_worker->state, task);
+
+		new_worker->state.has_task = true;
+		run_worker(new_worker);
+		return result;
+	}
+
+private:
+	vector<ReportableWorker*> threads;
+
+	static std::chrono::milliseconds THREAD_LIVE_TIMEOUT;
+	static size_t ETERNAL_THREAD_COUNT;
+	static size_t START_THREAD_COUNT;
```

Comment:
Constants may be expressed more explicitly like
`constexpr static size_t ETERNAL_THREAD_COUNT = 2;`

Reply:
Done